typegraph/python_next/typegraph_next/graph/typegraph.py

```diff
@@ -60,3 +43,81 @@ def expose(self, **kwargs: "t.func"):
         res = core.expose(store, lst, [])
         if isinstance(res, Err):
             raise Exception(res.value)
+
+
+@dataclass
+class G:
+    typegraph: Typegraph
+
+    def expose(self, **kwargs: "t.func"):
+        self.typegraph.expose(**kwargs)
+
+
+def __build_typegraph(
+    builder: Callable[[G], None],
+    name: Optional[str],
+    dynamic: Optional[bool] = None,
+    folder: Optional[str] = None,
+) -> Typegraph:
+    if name is None:
+        import re
+
+        # To kebab case
+        name = re.sub("_", "-", builder.__name__)
+
+    tg = Typegraph(name, dynamic, folder)
+    Typegraph._context.append(tg)
+    core.init_typegraph(
+        store,
+        TypegraphInitParams(
+            name=tg.name, dynamic=tg.dynamic, folder=tg.folder, path=tg.path
+        ),
+    )
+
+    builder(G(tg))
+
+    popped = Typegraph._context.pop()
+    assert tg == popped
+
+    res = core.finalize_typegraph(store)
+    if isinstance(res, Err):
+        raise Exception(res.value)
+
+    print(res.value)
+
+    return tg
+
+
+@overload
+def typegraph(builder: Callable[[G], None]) -> Typegraph:
+    ...
+
+
+@overload
+def typegraph(
+    name: str, dynamic: Optional[bool] = None, folder: Optional[str] = None
```

Comment:
simplify with typegraph(name, {dynamic, folder} = None)?

Reply:
I don't understand well.
Require keyword-argument??

Reply:
I was a bit too fast... if intelisense is confused, you could simplify with def typegraph(name: str, opts: Optional[Opts] = None) where Opts is typed dict with arguments?

Reply:
It is ok now. I removed the overload so when the name should be inferred and there is no additional argument, we would have to add the parenthesis:
```python
@typegraph()
def my_tg():
    # ...
```